cmd/goplugin-identity/identity/identity.go

```diff
@@ -1,199 +1,505 @@
 	filename string
 }
 
+// A tuple represents an external ID with timestamp and GC status
+type tuple struct {
+	InternalID string
+	ExternalID string
+	Timestamp  time.Time
+	Era        int64
+}
+
+type storeMeta struct {
+	Version    string
+	Timestamp  time.Time
+	Era        int64
+}
+
+
+// error used internally by the identity service
+type identityError string
+
+func errorf(f string, args ...interface{}) error {
+	return identityError(fmt.Sprintf(f, args...))
+}
+
+func (ie identityError) Error() string {
+	return string(ie)
+}
+
+var metadata = []byte("metadata")
+var internalToExternal = []byte("internalToExternal")
+var externalToInternal = []byte("externalToInternal")
+var garbage = []byte("garbage")
+
+var identityStoreVersion = semver.MustParseVersion("1.0.0")
+var supportedVersions = semver.MustParseVersionRange("1.x")
+
+// Start the Identity service running
+func Start(filename string) {
+	eval.Puppet.Do(func(c eval.Context) {
+		sb := service.NewServerBuilder(c, "Default::Identity::Service")
+		id, err := NewIdentity(filename)
+		if err != nil {
+			panic(err)
+		}
+		sb.RegisterAPI(serviceapi.IdentityName, id)
+		s := sb.Server()
+		grpc.Serve(c, s)
+	})
+}
+
+// ValueTuple creates a four element Array consisting of InternalID, ExternalID, Timestamp, and GCEra.
+//
+// The Pcore type of the tuple is Tuple[String, String, Timestamp, Integer]
+func (t *tuple) ValueTuple() eval.List {
+	return types.WrapValues([]eval.Value{
+		types.WrapString(t.InternalID),
+		types.WrapString(t.ExternalID),
+		types.WrapTimestamp(t.Timestamp),
+		types.WrapInteger(t.Era)})
+}
+
 // NewIdentity opens the database
 func NewIdentity(filename string) (*Identity, error) {
-	i := &Identity{
-		filename: filename,
-	}
-	db, err := i.openDB()
-	defer db.Close()
+	absName, err := filepath.Abs(filename)
 	if err != nil {
 		return nil, err
 	}
-	return i, nil
-}
-
-func (i *Identity) openDB() (*bolt.DB, error) {
-	db, err := bolt.Open(i.filename, 0600, &bolt.Options{Timeout: 200 * time.Millisecond})
-	if err != nil {
-		return nil, err
+	i := &Identity{
+		filename: absName,
 	}
-	err = db.Update(func(tx *bolt.Tx) error {
-		_, err := tx.CreateBucketIfNotExists([]byte("internalToExternal"))
-		if err != nil {
-			return err
-		}
-		_, err = tx.CreateBucketIfNotExists([]byte("externalToInternal"))
-		if err != nil {
+	err = i.withDb(func(db *bolt.DB) error {
+		// Ensure that buckets exist
+		return db.Update(func(tx *bolt.Tx) error {
+			mbb := tx.Bucket(metadata)
+			if mbb != nil {
+				mb := mbb.Get(metadata)
+				if mb == nil {
+					return fmt.Errorf("identity store at '%s' has invalid format", i.filename)
+				}
+				md := unmarshalMetadata(mb)
+				if !supportedVersions.Includes(semver.MustParseVersion(md.Version)) {
+					return fmt.Errorf("identity store at '%s' has unsupported data store version. Expected %s, got %s", i.filename, supportedVersions, md.Version)
+				}
+				return nil
+			}
+
+			// No metadata exists. May still be an older version
```

Comment:
No older identity DBs really exist.

Reply:
My thought was that someone who's testing our stuff, Dimitri, Kenaz, or someone else not involved with development, might encounter problems with a lingering identity.db file. Intention here is to highlight the problem.